src/graph/immutable_graph.cc

```diff
@@ -0,0 +1,975 @@
+  CHECK(IsValidIdArray(vids)) << "Invalid vertex id array.";
+  const auto len = vids->shape[0];
+  const dgl_id_t* vid_data = static_cast<dgl_id_t*>(vids->data);
+  DegreeArray rst = DegreeArray::Empty({len}, vids->dtype, vids->ctx);
+  dgl_id_t* rst_data = static_cast<dgl_id_t*>(rst->data);
+  for (int64_t i = 0; i < len; ++i) {
+    const auto vid = vid_data[i];
+    CHECK(HasVertex(vid)) << "Invalid vertex: " << vid;
+    rst_data[i] = this->GetDegree(vid);
+  }
+  return rst;
+}
+
+class Bitmap {
+  const size_t size = 1024 * 1024 * 4;
+  const size_t mask = size - 1;
+  std::vector<bool> map;
+
+  size_t hash(dgl_id_t id) const {
+    return id & mask;
+  }
+ public:
+  Bitmap(const dgl_id_t *vid_data, int64_t len): map(size) {
+    for (int64_t i = 0; i < len; ++i) {
+      map[hash(vid_data[i])] = 1;
+    }
+  }
+
+  bool test(dgl_id_t id) const {
+    return map[hash(id)];
+  }
+};
+
+/*
+ * This uses a hashtable to check if a node is in the given node list.
+ */
+class HashTableChecker {
+  std::unordered_map<dgl_id_t, dgl_id_t> oldv2newv;
+  // This bitmap is used as a bloom filter to remove some lookups.
+  // Hashtable is very slow. Using bloom filter can significantly speed up lookups.
+  Bitmap map;
+
+  void Collect(const dgl_id_t old_id, const dgl_id_t old_eid,
+               std::vector<dgl_id_t> *col_idx,
+               std::vector<dgl_id_t> *orig_eids) {
+    if (!map.test(old_id))
+      return;
+
+    auto it = oldv2newv.find(old_id);
+    if (it != oldv2newv.end()) {
+      const dgl_id_t new_id = it->second;
+      col_idx->push_back(new_id);
+      if (orig_eids)
+        orig_eids->push_back(old_eid);
+    }
+  }
+
+ public:
+  HashTableChecker(const dgl_id_t *vid_data, int64_t len): map(vid_data, len) {
+    oldv2newv.reserve(len);
+    for (int64_t i = 0; i < len; ++i) {
+      oldv2newv[vid_data[i]] = i;
+    }
+  }
+
+  void CollectOnRow(const dgl_id_t neigh_idx[], const dgl_id_t eids[], size_t row_len,
+                    std::vector<dgl_id_t> *new_neigh_idx,
+                    std::vector<dgl_id_t> *orig_eids) {
+    // TODO(zhengda) I need to make sure the column index in each row is sorted.
+    for (size_t j = 0; j < row_len; ++j) {
+      const dgl_id_t oldsucc = neigh_idx[j];
+      const dgl_id_t eid = eids[j];
+      Collect(oldsucc, eid, new_neigh_idx, orig_eids);
+    }
+  }
+};
+
+std::pair<ImmutableGraph::CSR::Ptr, IdArray> ImmutableGraph::CSR::VertexSubgraph(
+    IdArray vids) const {
+  const dgl_id_t* vid_data = static_cast<dgl_id_t*>(vids->data);
+  const int64_t len = vids->shape[0];
+
+  HashTableChecker def_check(vid_data, len);
+  // check if varr is sorted.
+  CHECK(std::is_sorted(vid_data, vid_data + len)) << "The input vertex list has to be sorted";
+
+  // Collect the non-zero entries in from the original graph.
+  std::vector<dgl_id_t> orig_edge_ids;
+  orig_edge_ids.reserve(len * 50);
+  auto sub_csr = std::make_shared<CSR>(len, len * 50);
```

Comment:
What is 50 here?

Reply:
this is the rough estimation the ration of #edges/#nodes. it's not very accurate.

Reply:
What if we have much fewer edges than nodes? Will this cause too much memory request for the allocator? The magic number is hard to maintain.

Reply:
we could assume #edges is the same as #vertices  and let it grow when necessary.